hal/ia32/acpi.c

```diff
@@ -0,0 +1,165 @@
+#include <hal/hal.h>
+
+#define MAX_CPU_COUNT 64
+
+typedef struct {
+	char magic[4];
+	u32 length;
+	u8 revision;
+	u8 checksum;
+	char oem[6];
+	char oemID[8];
+	u32 oemRevision;
+	u32 creatorId;
+	u32 creatorRevision;
+} __attribute__((packed)) sdt_header_t;
+
+
+typedef struct {
+	char magic[8];
+	u8 checksum;
+	char oemID[6];
+	u8 revision;
+	struct {
+		sdt_header_t header;
+		sdt_header_t *sdt[];
+	} __attribute__((packed)) * rsdt;
+} __attribute__((packed)) rsdp_t;
+
+
+typedef struct {
+	rsdp_t rsdp;
+
+	u32 length;
+	u64 xsdtAddr;
+	u8 extChecksum;
+	u8 _reserved[3];
+} __attribute__((packed)) xsdp_t;
+
+
+static inline int _hal_checksumVerify(const void *data, const size_t size)
+{
+	const u8 *ptr;
+	u8 sum = 0;
+	for (ptr = (const u8 *)data; ptr < (const u8 *)data + size; ++ptr) {
+		sum += *ptr;
+	}
+	return sum == 0;
+}
+
+
+static void _hal_acpiMadtHandler(hal_syspage_t *hs, sdt_header_t *header)
+{
+	if (_hal_checksumVerify(header, header->length) != 0) {
+		hs->madt = (unsigned int)header;
+		hs->madtLength = header->length;
+	}
+}
+
+
+static void _hal_acpiFadtHandler(hal_syspage_t *hs, sdt_header_t *header)
+{
+	if (_hal_checksumVerify(header, header->length) != 0) {
+		hs->fadt = (unsigned int)header;
+		hs->fadtLength = header->length;
+	}
+}
+
+static const struct {
+	const char magic[4];
+	void (*handler)(hal_syspage_t *hs, sdt_header_t *);
+} acpi_knownTables[] = {
+	{ .magic = "APIC", .handler = _hal_acpiMadtHandler },
+	{ .magic = "FACP", .handler = _hal_acpiFadtHandler }
+};
+
+
+static rsdp_t *_hal_acpiLookForRsdp(hal_syspage_t *hs, void *start, void *end)
+{
+	static const char RSDP_MAGIC[] = "RSD PTR ";
+	void *ptr;
+	rsdp_t *entry;
+	for (ptr = start; ptr < end; ptr += 16) {
+		if (hal_strncmp(RSDP_MAGIC, ptr, 8) == 0) {
+			if (_hal_checksumVerify(ptr, sizeof(rsdp_t)) != 0) {
+				entry = ptr;
+				if ((entry->revision == 2) && (_hal_checksumVerify(entry, sizeof(xsdp_t)) != 0)) {
+					hs->acpi_version = ACPI_XSDP;
```

Comment:
Where does the `ACPI_XSDP` come from? (looks like something hasn't been staged to commit)

Reply:
They're included outside of plo in: `phoenix-rtos-kernel/include/arch/syspage-ia32.h`